filters/openpolicyagent/openpolicyagent.go

```diff
@@ -514,10 +543,84 @@ func (opa *OpenPolicyAgentInstance) Start(ctx context.Context, timeout time.Dura
 	return nil
 }
 
+// Start starts the policy engine's plugin manager and periodically trigger the plugins to download policies etc.
+func (opa *OpenPolicyAgentInstance) StartAndTriggerPlugins(ctx context.Context) error {
+	err := opa.manager.Start(ctx)
+
+	if err == nil {
+		err = withRetry(ctx, opa.triggerPlugins, []time.Duration{100 * time.Millisecond, 200 * time.Millisecond}, opa.Logger(), "trigger bundles")
+	}
+	if err == nil {
+		err = opa.verifyAllPluginsStarted()
+	}
+
+	if err != nil {
+		opa.Close(ctx)
+		return err
+	}
+	return nil
+}
+
+func withRetry(ctx context.Context, httpCall func(ctx context.Context) error, waits []time.Duration, logger logging.Logger, operation string) error {
+	var httpError download.HTTPError
+
+	err := httpCall(ctx)
+	for i, backoff := range waits {
+		isHttpError := errors.As(err, &httpError)
+		if !isHttpError || httpError.StatusCode < 500 {
+			break
+		}
+		logger.WithFields(map[string]interface{}{
+			"retry":  i,
+			"status": httpError.StatusCode,
+		}).Warn("%v failed. Retrying %d", operation, i)
+
+		time.Sleep(backoff)
+		err = httpCall(ctx)
+	}
+	return err
+}
+
+func (opa *OpenPolicyAgentInstance) verifyAllPluginsStarted() error {
+	allPluginsReady := true
+	for pluginName, status := range opa.manager.PluginStatus() {
+		if status != nil && status.State != plugins.StateOK {
+			opa.Logger().WithFields(map[string]interface{}{
+				"plugin_name":   pluginName,
+				"plugin_state":  status.State,
+				"error_message": status.Message,
+			}).Error("Open policy agent plugin failed to start %s", pluginName)
+
+			allPluginsReady = false
+		}
+	}
+	if !allPluginsReady {
+		return fmt.Errorf("open policy agent plugins failed to start")
+	}
+	return nil
+}
+
+func (opa *OpenPolicyAgentInstance) triggerPlugins(ctx context.Context) error {
+
+	for _, pluginName := range []string{"discovery", "bundle"} {
+		if plugin, ok := opa.manager.Plugin(pluginName).(plugins.Triggerable); ok {
+			if err := plugin.Trigger(ctx); err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("plugin %s not found", pluginName)
```

Comment:
This would fail if there is no discovery config ?

Reply:
in the tests it does not but I will have a look, the behaviour might change

Reply:
make discovery plugin optional